pkg/kubelet/deviceplugin/types.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deviceplugin
+
+import (
+	"sync"
+
+	"google.golang.org/grpc"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
+)
+
+// MonitorCallback is the function called when a device's health state changes,
+// or new devices are reported, or old devices are deleted.
+// Updated contains the most recent state of the Device.
+type MonitorCallback func(resourceName string, added, updated, deleted []*pluginapi.Device)
+
+// Manager manages all the Device Plugins running on a node.
+type Manager interface {
+	// Start starts the gRPC Registration service.
+	Start() error
+	// Devices is the map of devices that have registered themselves
+	// against the manager.
+	// The map key is the ResourceName of the device plugins.
+	Devices() map[string][]*pluginapi.Device
+
+	// Allocate takes resourceName and list of device Ids, and calls the
+	// gRPC Allocate on the device plugin matching the resourceName.
+	Allocate(string, []string) (*pluginapi.AllocateResponse, error)
+
+	// Stop stops the manager.
+	Stop() error
+}
+
+// ManagerImpl is the structure in charge of managing Device Plugins.
+type ManagerImpl struct {
+	socketname string
+	socketdir  string
+
+	Endpoints map[string]*endpoint // Key is ResourceName
+	mutex     sync.Mutex
+
+	callback MonitorCallback
+
+	server *grpc.Server
+}
+
+const (
```

Comment:
`errDevicePluginAlreadyExists`, `errDevicePluginUnknown`, `errDeviceUnknown` are not used anywhere.

Every other error except for `errListAndWatch` is used at just one place. It would make these much more readable if you replace the use of these constants with the string directly at those places. I don't see the utility of these constants.

Reply:
Removed errDevicePluginUnknown and errDeviceUnknown. errDevicePluginAlreadyExists will be used later when we handle duplicate registrations with more intelligent logic.

For the error string definition, I also generally prefer to using them directly. But I don't know whether we need to define them if we want to monitor such errors through NPD. Lets keep them here for now in this PR. Added a TODO about possible cleanup later.